pkgs/misc/cups/drivers/foomatic-db-engine/default.nix

```diff
@@ -0,0 +1,81 @@
+{ lib
+, perlPackages
+, fetchFromGitHub
+, autoconf
+, automake
+, file
+, makeWrapper
+}:
+
+perlPackages.buildPerlPackage rec {
+  pname = "foomatic-db-engine";
+  version = "unstable-2022-05-03";
+
+  src = fetchFromGitHub {
+    # there is also a daily snapshot at the `downloadPage`,
+    # but it gets deleted quickly and would provoke 404 errors
+    owner = "OpenPrinting";
+    repo = "foomatic-db-engine";
+    rev = "2e6f14b54748fa121a4d2e3d480010e10b070c5a";
+    hash = "sha256-m7FQTxWmawbtm24h8UqznGKXgX41JhOtyyFMRwEhm5k=";
+  };
+
+  outputs = [ "out" ];
+
+  propagatedBuildInputs = [
+    perlPackages.Clone
+    perlPackages.DBI
+    perlPackages.XMLLibXML
+  ];
+  # The following packages would be referenced by
+  # the final package if available during build time,
+  # they might add or enhance some funcionality:
+  # cups cups-filters curl netcat-gnu samba wget
+
+  nativeBuildInputs = [ autoconf automake file makeWrapper ];
+  # The following package would be "found"
+  # by the configure script if available,
+  # it might improve something in the final package:
+  # a2ps ghostscript
```

Comment:
What do you mean by "improve something"? You noticed the configure script looks for these commands, but you didn't understand what's the usage of the configure script / make files with these commands?

Reply:
Somehow yes.  If you put `a2ps` into `nativeBuildInputs`, the configure script will say something like

```
checking for a2ps... /nix/store/fhp0m7anlcx4zpzz7qy0kkyilpag5j5n-a2ps-4.14/bin/a2ps
```

For `a2ps` specifically, I couln't find any usage inside the `foomatic-db-engine` sources, so it *might* be some kind of leftover, but I can't really be sure.

This is similar to the list I put in another comment a few lines above: There are also packages that are referenced by the configure script if found, but they will also become a runtime dependency of the resulting `foomatic-db-engine` package in the end (in contrast to `a2ps` and `ghostscript` which will be "used" during the build process, but not be referenced by the result).

The result works fine without any of those packages as build inputs.  On the other hand, the configure script demands certain build inputs (like the `file` utility), so I came to the conclusion that these packages are deliberately treated as optional inputs and can be omitted to avoid bloating the build and the closure.

However, someday someone might be missing a feature that could be unlocked with one of those optional input packages and I hope these comments might point that someone into the right direction and save some debugging time.

Reply:
`./configure` scripts looking up for `file` is a known harmless issue happening with all autotools based software, nothing I'd worry about.

OTH, `foomatic-db-engine`'s README explains what is the usage of the `a2ps` and ghostscript and it seems to me, that they can be used during runtime as filters or something like that. It's not a reason to add them to `nativeBuildInputs`. As for the `buildInputs`, it would have been nice to track down what are the new features gained by adding `cups`, `cups-filter` etc.

If I understand correctly, that most of `foomatic-db-engine`'s purpose is to help build `foomatic-db-ppds`, I trust you that most users will likely use the ppds and that the generated ppds are good. Still, perhaps it'd be best to add these inputs conditionally, defaulting to not adding them; and ideally, explain better in a comment above the `lib.optional` lines what is the feature gained by adding these dependencies. Perhaps permalinks to upstream's README is best.

Reply:
Sadly, the README file didn't really help here.  It recommends to have a couple of additional programs installed, but some of those are not really used by the engine package (probably just recommended to make certain PPD files work).  Other dependencies are not mentioned in the README.

I tracked down the references in the perl code: `ghostscript` and `a2ps` are not used at all; they do not alter the output of the package in any way.  The other dependencies expand the functionalities of the scripts `foomatic-{configure,getpjloptions,printjob}`.  From what I understand, these script can be used to modify printer configuration (e.g. add CUPS queues) or submit print jobs, so they are indeed not used in the process of PPD file generation.

I added three switches `withCupsAccess`, `withSocketAccess` and `withSMBAccess` that control the addition of optional dependencies, as you suggested.  I was thinking about the common name `with...Support` that is used elsewhere in nixpkgs, but concluded it might confuse other contributors into thinking that `foomatic-db-engine` or the PPD files it generates cannot be used with cups unless `withCupsSupport` is enabled.  The name `with...Access` makes it more clear that this is only needed if the resulting package is going to be used to *access* one of those entites directly.

Those new switches are the result of my interpretation of the perl code (in particular, the `sysdeps` structure created by `makeDefaults.pm`).  I'm quite sure that access to cups, to network socket and to SMB can be separated the way presented here.  It might even be possible to divide the cups access feature up further in a meaningful way, but from looking at the code, I can't spot an obvious way to do this.

Reply:
Very good job on `foomatic-db-engine`! Thanks again!